tests/integration/valkey-cli.tcl

```diff
@@ -611,41 +611,65 @@ if {!$::tls} { ;# fake_redis_node doesn't support TLS
     test {valkey-cli pubsub mode with multiple subscription types} {
         set fd [open_cli]
         
-        # Subscribe to a regular channel
-        write_cli $fd "SUBSCRIBE channel1"
-        assert_match "*subscribe*channel1*" [read_cli $fd]
-
-        # Subscribe to a pattern
+        write_cli $fd "SUBSCRIBE channel1 channel2 channel3"
+        set response [read_cli $fd]
+
+        foreach line [split $response "\n"] {
+            if {[string match "*subscribe*channel1*" $line]} {
+                assert_match "*subscribe*channel1*" $line
+            } elseif {[string match "*subscribe*channel2*" $line]} {
+                assert_match "*subscribe*channel2*" $line
+            } elseif {[string match "*subscribe*channel3*" $line]} {
+                assert_match "*subscribe*channel3*" $line
+            }
+        }
+        
         write_cli $fd "PSUBSCRIBE pattern*"
         assert_match "*psubscribe*pattern**" [read_cli $fd]
 
-        # Subscribe to a shard channel
         write_cli $fd "SSUBSCRIBE schannel1"
         assert_match "*ssubscribe*schannel1*" [read_cli $fd]
 
-        # Unsubscribe from regular channel
         write_cli $fd "UNSUBSCRIBE channel1"
         assert_match "*unsubscribe*channel1*" [read_cli $fd]
 
         # Verify still in pubsub mode
         catch {run_command $fd "SET key value"} err
         assert_match "*ERR*only*SUBSCRIBE*UNSUBSCRIBE*allowed*" $err
 
-        # Unsubscribe from pattern
         write_cli $fd "PUNSUBSCRIBE pattern*"
         assert_match "*punsubscribe*pattern**" [read_cli $fd]
 
-         # Verify still in pubsub mode
+        # Verify still in pubsub mode
         catch {run_command $fd "GET key"} err
         assert_match "*ERR*only*SUBSCRIBE*UNSUBSCRIBE*allowed*" $err
 
-        # Unsubscribe from shard channel
         write_cli $fd "SUNSUBSCRIBE schannel1"
         assert_match "*sunsubscribe*schannel1*" [read_cli $fd]
 
-        # Verify that we've exited pubsub mode
-        assert_equal "PONG" [run_command $fd "PING"]
+        catch {run_command $fd "INCR counter"} err
+        assert_match "*ERR*only*SUBSCRIBE*UNSUBSCRIBE*allowed*" $err
+
+        # Unsubscribe from all remaining channels
+        write_cli $fd "UNSUBSCRIBE"
 
+        set combined_response [read_cli $fd]
+
+        foreach line [split $combined_response "\n"] {
+            if {[string match "*unsubscribe*channel2*" $line]} {
+                assert_match "*unsubscribe*channel2*" $line
+            } elseif {[string match "*unsubscribe*channel3*" $line]} {
+                assert_match "*unsubscribe*channel3*" $line
+            }
+        }
+
+        # Verify that we've exited pubsub mode
+        set response [run_command $fd "PING"]
+        
+        if {[string first "(subscribed mode)" $response] >= 0} {
+            return -code error "Client is still in pubsub mode"
+        }
```

Comment:
Interesting! Does this work though? Does the response really contain the prompt?

We should add the opposite check when we enter subscribed mode, to see that the prompt does include "(subscribed mode)" when we are in subscribed mode. In this way, we also know that this kind of check works.

Reply:
I've done some source code reading. The prompt is only printed when STDIN is a TTY. When we run the cli from the tests, the prompt is not printed.

(The code for this is in the command line editor linenoise (`deps/linenoise/linenoise.c`), function `linenoise()`, which calls linenoiseNoTTY which doesn't print the prompt.)

So, to test if valkey-cli thinks that it's in pubsub mode or not, we need to do something else. Either we can modify linenoise to add some way to make it behave like an interactive terminal even it it's not running in a terminal. That seems complicated though. I have another idea.

There is some prepared work in `valkey-cli` for "internal commands", starting with ":".

```
not connected> :banana
unknown valkey-cli internal command ':banana'
```

The only such commands right now are `:set nohints` and `:set hints` to disable or enable syntax hints.

We could add an internal command to be able to check if we're in pubsub mode. I think for example `:get prompt` (to print the `config.prompt`) would be simple enough and useful for testing. WDYT?